src/commands.cc

```diff
@@ -325,6 +326,23 @@ void write_buffer(const ParametersParser& parser, Context& context, const ShellC
     auto filename = parser.positional_count() == 0 ?
                         buffer.name() : parse_filename(parser[0]);
 
+    struct stat st;
+    auto restore_mode = on_scope_end([&]{
```

Comment:
This should not be done before the call to `::stat`, imagine if `::stat` fails, then we will be doing a `::chmod` using an uninitialized `st`.